Updated Readme.md with some definitions and tables. (#3)
finntonwentworth/typogenetics: README.md

| [mvl](#mvl)  | Move one unit to the left | |
| [cop](#cop)  | Turn on copy mode | |
| [off](#off)  | Turn off copy mode | |
| [ina](#ina)  | Insert A to the right of this unit | |
| [inc](#inc)  | Insert C to the right of this unit | |
| [ing](#ing)  | Insert G to the right of this unit | |
| [int](#int)  | Insert T to the right of this unit | |
| [rpy](#rpy)  | Search of the nearest pyrimidine to the right | |
| [rpu](#rpu)  | Search for the nearest purine to the right | |
| [lpy](#lpy)  | Search for the nearest pyrimidine to the left | |
| [lpu](#lpu)  | Search for the nearest purine to the left   | |

#### pun

Not an Amino acid/instruction, but deserves special clarification. 

Punctuates the strand, allowing one strand to encode for two or more separate enzymes. 
When determining the enzymes encoded for by a strand, the prescence of the pun instruction 
ends the enzyme early, and an additional enzyme is determined from the remaining strand, decoding 
until the end of the strand or another pun is reached. These enzymes will act sequentially on
the strand, one at a time. The portion of a strand that encodes for a specific enzyme is known

> as a 'gene', so one strand may contain one or more genes that encode for the same number of enzymes. 

Reviewer: > Isn't it possible for a strand to have zero genes if it only consists of `AA`?

Yes, that's correct. I misunderstood the idea behind the AA base pair. It's not an amino acid, and it does not produce an enzyme. 

> Also, if a strand produces two enzymes, do they really process the entire strand? Not just the gene they were created from?

This is actually ambiguous, since no demonstration of 2 separate enzymes is shown. That was my big change in understanding behind punctuation when I first start implementing it. See [Typogenetics](https://citeseerx.ist.psu.edu/document?repid=rep1&type=pdf&doi=a512a815e6d7cd8a3c87c9997802ec35f0505bf7) , page 10. (As an aside, I have not viewed any of the code or examples in that paper, just found that looking for rules clarifications). So we should mention that this is the execution we chose in the Authors' notes.

Pun has no folding direction, and therefore does not contribute to the folding pattern of the enzyme, 
as well as no additional effects on the enzyme past encoding. 
 
#### cut

Cuts the strand at the unit the enzyme is currently bound to.
The cut is placed to the right of the bound unit.
Cut applies to both the "active" strand and the "complementary" strand that
is being generated from the copy mode.

#### del 

Deletes the base that the enzyme is attached to, and then the enzyme moves to the right. 

> Only pertains to the active strand, *not* the complementary strand. 

Reviewer: > Actually, after reading mvl and mlr I'm assuming it does not "reconnect" the bases after deleting a base. Does this mean del will result in two strands?

Yes, del leaves a blank space that the enzyme can fall off of. If you del a base, then mvl, you'd end the enzyme acting. And we should mention that it is not affected by copy mode, that's a good point

#### swi

Switches the enzyme from the "active" strand to the "complementary" strand. Since the complementary strand
is represented upside-down above the active strand (see [cop](#cop)), switching strands seems to reverse the order 
of the bases in the now active strand. If a switch occurs and there are no bases in the complementary strand above 
where the enzyme is acting, the enzyme falls off and ends acting.  

#### mvr

Moves the enzyme one base to the right. If the enzyme moves off the end of the strand or into a gap with no bases, then it finishes executing
even if it had more instructions left to execute. 

#### mvl

Moves the enzyme one base to the left. If the enzyme moves off the front of the strand or into a gap with no bases, then it finishes executing
even if it had more instructions left to execute, same as in mvr. 

#### cop

Places the complement of the current base into the complementary strand, and enters "copy mode." Each base has its corresponding

> complement, known as the complementary base pairin (see [Definitions](#Definitions)). Copy mode remains in effect until an off 

Author: ```suggestion
complement, known as the complementary base pairing (see [Definitions](#Definitions)). Copy mode remains in effect until an [off](#off) 
```

instruction is reached (or the end of the strand/enzyme). While in copy mode, any base that is "touched" by the enzyme will 
have its complementary base placed above it in the complementary strand. When bases are copied into the complementary 
strand, they are placed *upside-down* above the element copied, so that the strands are mirrored to each other. Copy mode applies 

> after a move, so that the movement off of a strand or into a gap between bases still preserves the behavior of the strand ending normally.  

Author: ```suggestion
*after* a move, so that the movement off of a strand or into a gap between bases still preserves the behavior of the strand ending normally.  
```
I think extra emphasis on the fact copy mode is applied after is important.

While in copy mode, additional cop instructions do nothing. 

#### off

Turns off copy mode. If copy mode is not enabled, then this does nothing. 

#### ina

Inserts an 'A' base into the position to the right of the enzyme, and if copy mode is on, then the base's complement (T) is placed above into

> the complementary strand. 

Author: Insert commands don't cause the enzyme to move to the right correct? So if you had `ina inc int ing` you would get `GTCA`? Since it always inserts to the right

Reviewer: Yes that is correct. For some reason I was under the impression that all instructions moved the enzyme one unit to the right after they were done executing.

Author: Funny enough, since enzymes aren't required to always bind to the first binding base, you could have enzymes that primarily move from left to right instead of left to right.

It would be interesting to see if we could find a palindrome strand that would do the same thing backwards and forwards, depending on if it binds to the left or right side first.

Perhaps a 'select last binding base' flag would be interesting.

the complementary strand. 

#### rpy

Moves to the nearest pyrine to the right of this unit. (See [Definitions](#Definitions)). With copy mode enable, each base touched while sliding to the next pyrine 
will be copied into the complementary strand. If there are no pyrines to the right of the current base, this does nothing.

#### rpu

Moves to the nearest purine to the right of this unit. (See [Definitions](#Definitions)). With copy mode enable, each base touched while sliding to the next purine 
will be copied into the complementary strand. If there are no purines to the right of the current base, this does nothing.

#### lpy

Moves to the nearest pyrine to the left of this unit. (See [Definitions](#Definitions)). With copy mode enable, each base touched while sliding to the next pyrine 
will be copied into the complementary strand. If there are no pyrines to the left of the current base, this does nothing.

#### lpu

Moves to the nearest purine to the left of this unit. (See [Definitions](#Definitions)). With copy mode enable, each base touched while sliding to the next purine 

> will be copied into the complementary strand. If there are no purines to the left of the current base, this does nothing.

Author: Maybe move the `(See [Definitions](#Definitions))` bit to be directly after the pyrine/purine word?

Author: Also it should probably be `With copy mode enabled,...` instead of `With copy mode enable,...`.

Author: In the case that there is no purines/pyrines to the left/right, does it actually do nothing? Or does it try to go that direction until it falls off? Because if copy mode is on then it is filling in all those empty complements till the end of the strand.

Reviewer: > In the case that there is no purines/pyrines to the left/right, does it actually do nothing? Or does it try to go that direction until it falls off? Because if copy mode is on then it is filling in all those empty complements till the end of the strand.

Another ambiguous point, so we should make an implementation decision. Secondarily, we have to decide if these instructions are capable of traversing gaps in strands. 
I would argue that the behavior is more interesting if they do nothing if there are no left/right purines/pyrines, as then there will be more strands that do not terminate prematurely in the space of all possible strands. I also think that they should not traverse gaps on the strand that they act on, as regular moves cannot traverse gaps either.

Author: Talked with Finn. We're going to make it so that these instructions dont have a look ahead behavior.

> ## Binding Preference

Author: What happens if a strand does not contain the base of an enzymes binding preference? It just doesn't act? Does that mean you can create strands that just won't operate on themselves and never will no matter how many iterations you run them for?

Reviewer: Yes that is exactly correct, and how it is implemented. If an enzyme does not have an available base to bind to, then it will not act on the strand. It prints with an exit message too. There are a large number of 'dud' strands actually, that do nothing or simply loop behavior. 
i.e. AAAA, AGGG

| &rarr; | &darr; | G |
| &rarr; | &larr; | T |

### Authors' Note: 
A number of ambiguities present in the original program are given defined behavior in this rendition of the system, listed below. 

> **Binding Preference of Single Amino Acid Enzymes**

Author: I thought this rule was for single base strands? Not single amino acid enzymes. Unless I'm misunderstanding something. Because if there was an enzyme with one amino acid, its first amino acid is also its last one, so the relative orientation is always the same direction, so it would always bind to A.

Reviewer: Yes, the clarification is for single amino acid strands. They do not have behavior as written, unless you allow that the first element in the folding pattern is also the last element. (which is how it is implemented currently). Single base strands do nothing and have no binding preference because they do not encode for any amino acids

> Since the binding preference of an enzyme is determined by the orientation of its first and last element, enzymes which consist of only a single instruction don't have a clearly defined 

Author: ```suggestion
Since the binding preference of an enzyme is determined by the relative orientation of its first and last amino acids, enzymes which consist of only a single amino acid don't have a clearly defined 
```

Author: Unsure about these edits. See above comment.

Reviewer: I think it's a good change, more correct language use.

rule for determining their starting base binding preference. We have elected to follow the fairly intuitive convention of having this class of enzymes start on 'A', counting their only element
as both the first and last fold. With both directions facing right, it follows that the enzyme would bind to base 'A' to begin with. 

**Starting Base Position Preference of Enzymes**

Hofstadter also fails to define which element the enzyme prefers to bind to when given a selection of options. Examples in the book show the first matching element most commonly selected, but 
this is not an explicitly defined rule. As such, the user is given a selection of bases to choose from when there are multiple matching bases in the strand. The user may also pass 2 parameters to

> the program, --Select-random (-s) and --first-matching (-f) which will have enzymes either randomly choose a starting base to begin on each time, or the first matching element of their starting base 

Author: Lowercase `--select-random`?

Author: Also maybe create a new flag or change the `--first-matching` flag to take a number of which starting option to take. Like, always select the n'th option. `--bind-position n`. First matching would then be `--bind-position 1`

Reviewer: > Lowercase `--select-random`?

I need to change the select base flags, they're currently -Sf or -Sr for 'Select first' or 'Select random'. Also there is no long form implementing yet, so entering --Select-random would just throw errors. 

> Also maybe create a new flag or change the `--first-matching` flag to take a number of which starting option to take. Like, always select the n'th option. `--bind-position n`. First matching would then be `--bind-position 1`

I was thinking about this, but it would be more akin to semi randomness. Because each time there is a choice for enzyme binding, the position numbers 1854000changes. So if the enzyme binds to 'A', that might correspond to Positions 1, 5, 10 in the strand. But the next enzyme that binds to 'C' won't have those options, instead they'll have 3,4,6 or whatever. So --bind-position would select array elements instead, and then there really isn't a choice there since it hard to predict what options will be available.
